x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/HoistRemoteEnrichTopN.java

```diff
@@ -0,0 +1,168 @@
+ * Then we need to add Project to remove the alias fields. Fortunately, PushDownUtils has the logic to do that.
+ */
+public final class HoistRemoteEnrichTopN extends OptimizerRules.OptimizerRule<Enrich> implements OptimizerRules.CoordinatorOnly {
+    public HoistRemoteEnrichTopN() {
+        super(OptimizerRules.TransformDirection.UP);
+    }
+
+    @Override
+    protected LogicalPlan rule(Enrich en) {
+        if (en.mode() == Enrich.Mode.REMOTE) {
+            LogicalPlan plan = en.child();
+            List<Attribute> outputs = en.output();
+            // This loop only takes care of one TopN. Repeated TopNs may be a problem, we can't handle them here.
+            while (true) {
+                if (plan instanceof TopN topN && topN.local() == false) {
+                    Set<Attribute> missingAttributes = checkOrderInOutputs(topN.order(), outputs);
+
+                    // Do we need to do any aliasing because some of the fields were overridden by Enrich or other generating plans?
+                    if (missingAttributes.isEmpty()) {
+                        // No need for aliasing - then it's simple, just copy the TopN on top of Enrich and mark the original as local
+                        LogicalPlan transformedEnrich = en.transformDown(TopN.class, t -> t == topN ? topN.withLocal(true) : t);
+                        return new TopN(topN.source(), transformedEnrich, topN.order(), topN.limit(), false);
+                    } else {
+                        Set<String> missingNames = new HashSet<>(Expressions.names(missingAttributes));
+                        AttributeMap.Builder<Alias> aliasesForReplacedAttributesBuilder = AttributeMap.builder();
+                        List<Expression> rewrittenExpressions = new ArrayList<>();
+
+                        // Replace the missing attributes in order expressions with their aliases
+                        for (Order expr : topN.order()) {
+                            rewrittenExpressions.add(expr.transformUp(Attribute.class, attr -> {
+                                if (missingNames.contains(attr.name())) {
+                                    Alias renamedAttribute = aliasesForReplacedAttributesBuilder.computeIfAbsent(attr, a -> {
+                                        String tempName = TemporaryNameUtils.locallyUniqueTemporaryName(a.name());
+                                        return new Alias(a.source(), tempName, a, null, true);
+                                    });
+                                    return renamedAttribute.toAttribute();
+                                }
+
+                                return attr;
+                            }));
+                        }
+                        var replacedAttributes = aliasesForReplacedAttributesBuilder.build();
+                        // We need to create a copy of attributes used by TonP, because Enrich or some command on the way overwrites them.
```

Comment:
```suggestion
                        // We need to create a copy of attributes used by TopN, because Enrich or some command on the way overwrites them.
```